src/runtime/request_manager.cc

```diff
@@ -378,131 +453,175 @@ BatchConfig RequestManager::prepare_next_batch(BatchConfig const &old_bc,
   // Step 2: prepare the next batch for existing requests
   BatchConfig new_bc;
   for (int i = 0; i < BatchConfig::max_requests_per_batch(); i++) {
-    if (old_bc.request_completed[i]) { // add new requests to the next batch
+    if (old_bc.request_completed[i]) {
+      // no need to carry over tokens to new batch for this request
       continue;
     } else {
       assert(old_bc.requestsInfo[i].num_tokens_in_batch > 0);
       Request &request = all_requests[old_bc.requestsInfo[i].request_guid];
-      int processed_tokens =
-          old_bc.requestsInfo[i].first_token_depth_in_request +
-          old_bc.requestsInfo[i].num_tokens_in_batch;
-      assert(processed_tokens < request.tokens.size());
-      bool request_completed = false;
-      // printf("model_type = %d\n", this->model_type);
-      if (request.tokens.size() >= old_bc.requestsInfo[i].max_sequence_length) {
-        request_completed = true;
-      } else if (request.tokens.back() == eos_token_id) {
-        // Encounter EOS token id
-        request_completed = true;
-      }
-      if (request_completed) {
-        std::string output = this->tokenizer_->Decode(request.tokens);
-        // Unlike Huggingface, the sentencepiece C++ library automatically
-        // removes the BOS token
-        if (model_type == ModelType::LLAMA &&
-            request.tokens.at(0) == bos_token_id) {
-          output = "<s> " + output;
-        }
-        {
-          // update generation result
+
+      if (request.req_type == RequestType::REQ_FINETUNING) {
+        // fine-tuning requests don't automatically carry over to the next
+        // batch, we only do so if there is space left after adding new
+        // inference requests
```

Comment:
Based on this comment, and based on how step 3 of `prepare_next_batch` is done (especially its condition to check whether index `i` is free space in the batch), I think `new_bc.request_completed[i]` should be set to `true` here.
Right now it should work without, because `new_bc.request_completed[i]` is not initialized so has a garbage value which has low probability of being 0, and so the condition `if(new_bc.request_completed[i])` in step 3 is evalutated (with high prob) to `true` already. But for consistency I think `new_bc.request_completed[i] = true` should be added here.

Reply:
Good catch! Let me make the fix!